backupccl: remap all restored tables

dt opened the pull request:
This PR has a few changes, broken down into separate commits:
a) stop restoring tmp tables and remove the special-case code to synthesize their special schemas; These were previously restored only to be dropped so that restored jobs that referenced them would not be broken, but we stopped restoring jobs.
b) synthesize type-change jobs during cluster restore; this goes with not restoring jobs.
c) fix some assumptions in tests/other code about what IDs restored tables have.
d) finally, always assign new IDs to all restored objects, even during cluster restore, removing the need to carefully move conflicting tables or other things around.

Commit-by-commit review recommended.

dt committed:
backupccl: synthesize type-change jobs during cluster restore

Release note: none.

dt committed:
backupccl: stop restoring temp tables

Release note: none.

dt committed:
backupccl: use less brittle method to find tmp system ID

Release note: none.

dt committed:
backupccl: make some tests less specific on IDs

Release note: none.

dt committed:
backupccl: remap IDs during cluster restore

Previously we kept the IDs of all descriptors being restored unchanged
during full-cluster restore. This had the advantage of meaning that all
mentions of IDs in anything we restored, for example in system tables or
jobs, were still valid as we put the same descriptor back at the same ID
when we restored it. However, this approach proved very brittle to
changes in what IDs we use for what over time; for example, in 22.1, new
system tables were introduced with IDs in the lower 50s, while those IDs
had previously been available to user-created tables which could appear
in the backup being restored. This posed a problem: we needed to restore
that backed up user table to ID 52, but there was a new system table in
the way. A workaround in 22.1 is that, mid-restore, we can pick up and
move an existing system table in the restoring cluster to a new ID,
picked to be greater than any ID we are going to restore into, then
carry on with the restore into the now conflict-free ID spans. However
this approach is brittle, as the pick-up-and-move could break any number
of assumptions or edges between descriptors. Fortunately for 22.1, no
such edges exist and only a couple new system descriptors had found
their way into the formerly user-owned IDs, however if left unchecked,
this problem could easily grow in complexity, with some future
descriptors being much harder to move.

Instead, this changes how we restore during full cluster restore to
match how we restore in other restores: each descriptor being restored
is assigned a new descriptor ID, generated as the next available ID in
the restoring cluster, and references to the old ID in other data being
restored are updated. Descriptor ID rewriting within descriptors
themselves has been in place as long as restore has existed, however
during full-cluster restore, data in addition to descriptors is
restored, particularly in system table rows. These rows can also contain
references to descriptor IDs, which is why a prior change started
updating these rows as they are restored to contain the new IDs. The
'old' and 'new' IDs were always the same in that change however, as
cluster-restore did not actually generate and assign _new_ IDs. That
last part is what this change does: instead of assigning the same ID, it
assigns a new one, and the rewriting enabled in prior changes is now
actually updating IDs.

The special-case moving of conflicting descriptors is no longer needed
now that all descriptors are assigned new IDs and is removed.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85492)
<!-- Reviewable:end -->

dt committed:
catprivilege: remove special cases for cluster restore

Release note: none.

benbardin reviewed (approved):
LGTM, but consider also waiting for another review or two from the others :)

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though the test failure seems plausibly legit. This was satisfying to read. 

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 1 of 3 files at r5, 4 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85492#-:-N8ZY7juBo3JHdmO-AQR:b-dngmjt)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru and @dt)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85492#-N8ZR2YZAywAtTmDwUZu):*
I'm so curious
___
*[`pkg/ccl/backupccl/restore_job.go` line 1667 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85492#-N8ZX2HNA3B18t2hRfKj:-N8ZX2HNA3B18t2hRfKk:bggeg8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f71d99b233bc547a40e91f1791609a247439d8f/pkg/ccl/backupccl/restore_job.go#L1667)):*
> ```Go
> 
> // tempSystemDatabaseID returns the ID of the descriptor for the temporary
> // system database used in full cluster restores.
> ```

Can you note that this returns `descpb.InvalidID` in the case that there are no system tables in the `tables` slice and that it relies on the `DescriptorRewrites` map having already been populated?
___
*[`pkg/ccl/backupccl/restore_old_versions_test.go` line 543 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85492#-N8ZXVjI1-oNp2IoDp09:-N8ZXVjI1-oNp2IoDp0A:bk00v3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/72c42549c47c01cab84a3991d4b3450bb82de5dc/pkg/ccl/backupccl/restore_old_versions_test.go#L543)):*
> ```Go
> 			{"root", "", "{admin}"},
> 		})
> 		sqlDB.CheckQueryResults(t, "SELECT comment FROM system.comments", [][]string{
> ```

technically, should this have an order by?


<!-- Sent from Reviewable.io -->

adityamaru commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -1665,21 +1665,22 @@ func (r *restoreResumer) notifyStatsRefresherOfNewTables() {
 
 // tempSystemDatabaseID returns the ID of the descriptor for the temporary
```
nit: can you add some more words tying this to the logic in planning. It will help future readers to understand why its okay to just use the first system tables parentID rewrite.

adityamaru commented on pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts:
```diff
@@ -69,56 +69,38 @@ SELECT count(1) FROM [SHOW TABLES FROM system] WHERE table_name LIKE 'crdb_inter
-# The system table IDs should have been replaced by restored descriptors.
 query-sql
-SELECT name, id FROM system.namespace WHERE id IN (104,105,106,107);
+SELECT name, id FROM system.namespace WHERE name LIKE 'crdb_internal_copy_%';
```
any reason to keep this statement around? Seems redundant with the previous one.

dt replied:
I figured we want to test both directions on the namespace table since the conflicts bug created duplicate entries and all

adityamaru commented on pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts:
```diff
@@ -153,4 +135,3 @@ should survive
 query-sql
 select role_name, settings FROM system.database_role_settings AS drs JOIN system.namespace AS ns ON ns.id = drs.database_id WHERE ns.name = 'foo';
 ----
-hamburger {application_name=helper}
```
hmm why don't we restore this anymore?

dt replied:
Ha, nice catch -- there was a bug in the prior PR that this was pointing to that was previously hidden by the fact we remapped 1:1. Fixed with new first commit here.

dt replied:
er, by first I meant 5th: [backupccl: remap all IDs in one pass](https://github.com/cockroachdb/cockroach/pull/85492/commits/6c54e9a8e6430c0f3b496d0909ae86f10920d50d)

stevendanna commented on pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts:
```diff
@@ -153,4 +135,3 @@ should survive
 query-sql
 select role_name, settings FROM system.database_role_settings AS drs JOIN system.namespace AS ns ON ns.id = drs.database_id WHERE ns.name = 'foo';
 ----
-hamburger {application_name=helper}
```
Is this expected?

dt replied:
No, see reply to Aditya above.

adityamaru reviewed (changes requested):
Can I bug you to flesh out the commit message in commit number 5. For anyone who comes along later it would be helpful to read why we made this switch etc. Feel free to ignore.

stevendanna reviewed (commented):
Overall, it is nice to see this code get deleted.

One question I have is why we don't need the custom restore funcs for comments, session vars, and zone configuration all of which reference descriptor IDs in their table.  If we _do_ need them, then perhaps the descriptor-conflicts tests isn't testing exactly what it used to be.

dt commented:
@stevendanna 
> One question I have is why we don't need the custom restore funcs for comments, session vars, and zone configuration all of which reference descriptor IDs in their table.
This is why https://github.com/cockroachdb/cockroach/pull/85098/files landed first

stevendanna commented:
@dt Ah, thanks. missed that one.

dt commented on pkg/ccl/backupccl/testdata/backup-restore/temp-tables:
```diff
@@ -146,14 +145,12 @@ pg_catalog
-temp_seq
-temp_table
 
 # Nudge and wait for the temp cleanup job to run.
```
It seemed like it was showing that we didn't have a tmp table post-restore, so that was worth keeping, but I could be convinced it isn't. It can also go away later?

dt replied:
Done.

dt committed:
backupccl: remap all IDs in one pass

Previously we updated one old ID to one new ID at a time. However this was
incorrect: if a new ID can potentially be higher or lower and conflict with
an old ID yet to be remapped, and thus subsequently be remapped again incorrectly
as it then looks like the other old ID. Since IDs can be higher or lower there
is no order that is sure to be free of such conflicts when doing each remapping
separately. Instead, we now do all the remappings in one pass, using a CASE to
switch from old ID to new ID as we go.

Release note: none.

dt commented on pkg/ccl/backupccl/testdata/backup-restore/temp-tables:
```diff
@@ -132,7 +132,6 @@ crdb_internal
 information_schema
 pg_catalog
 pg_extension
-pg_temp_0_0
```
Done.

dt commented:
> Can I bug you to flesh out the commit message in commit number 5. For anyone who comes along later it would be helpful to read why we made this switch etc. Feel free to ignore.

Added a lengthy commit message.

adityamaru commented on pkg/ccl/backupccl/system_schema.go:
```diff
@@ -478,15 +479,17 @@ func rekeySystemTable(
 		}
-		return nil
+		fmt.Fprintf(&q, "ELSE %s END", colName)
+		log.Warningf(ctx, "%s", q.String())
```
detritus?

dt commented:
TFTRs!

bors r+

dt commented:
bors r-

dt commented:
bors r+
